aioshelly/rpc_device.py

```diff
@@ -0,0 +1,202 @@
+        self.shelly = None
+        self._status: Dict[str, Any] | None = None
+        self._event: Dict[str, Any] | None = None
+        self._device_info = None
+        self._config = None
+        self._wsrpc = WsRPC(options.ip_address, self._on_notification)
+        self._update_listener = None
+        self._initialized = False
+        self._initializing = False
+
+    @classmethod
+    async def create(
+        cls,
+        aiohttp_session: aiohttp.ClientSession,
+        options: ConnectionOptions,
+        initialize: bool = True,
+    ):
+        """Device creation."""
+        instance = cls(aiohttp_session, options)
+
+        if initialize:
+            await instance.initialize()
+
+        return instance
+
+    async def _on_notification(self, method, params):
+        if method == "NotifyStatus" and params is not None:
+            self._status = dict(mergedicts(self._status, params))
+        elif method == "NotifyEvent" and params is not None:
+            self._event = params
+
+        if self._update_listener:
+            self._update_listener(self)
+
+    @property
+    def ip_address(self):
+        """Device ip address."""
+        return self.options.ip_address
+
+    async def initialize(self):
+        """Device initialization."""
+        self._initializing = True
+        self._initialized = False
+        try:
+            await self._wsrpc.connect(self.aiohttp_session)
+            self.shelly = await get_info(self.aiohttp_session, self.options.ip_address)
+
+            if self.options.auth or not self.shelly["auth_en"]:
+                await self.update_device_info()
+                await self.update_config()
+                await self.update_status()
+
+            self._initialized = True
+        finally:
+            self._initializing = False
+
+        if self._update_listener:
+            self._update_listener(self)
+
+    async def shutdown(self):
+        """Shutdown device."""
+        self._update_listener = None
+        await self._wsrpc.disconnect()
+
+    def subscribe_updates(self, update_listener):
+        """Subscribe to device status updates."""
+        self._update_listener = update_listener
+
+    async def update_status(self):
+        """Get device status from 'Shelly.GetStatus'."""
+        self._status = await self._wsrpc.call("Shelly.GetStatus")
+
+    async def update_device_info(self):
+        """Get device info from 'Shelly.GetDeviceInfo'."""
+        self._device_info = await self._wsrpc.call("Shelly.GetDeviceInfo")
+
+    async def update_config(self):
+        """Get device config from 'Shelly.GetConfig'."""
+        self._config = await self._wsrpc.call("Shelly.GetConfig")
+
+    @property
+    def initialized(self):
+        """Device initialized."""
+        return self._initialized
+
+    @property
+    def requires_auth(self):
+        """Device check for authentication."""
+        return self.shelly["auth_en"]
+
+    async def set_state(self, method, params):
```

Comment:
Why is this called `set_state` and not call_rpc or similar?

Reply:
Changed to `call_rpc`